File: src/main/java/olivine/FMap.java
package olivine;

public final class FMap {
  private final FMap parent;
  private final Term key, value;
  public static final FMap EMPTY = new FMap(null, null, null);

  public FMap add(Term key, Term value) {
    assert key.size() == 0;
    return new FMap(this, key, value);
  }

  private FMap(FMap parent, Term key, Term value) {
    this.parent = parent;
    this.key = key;
    this.value = value;
  }

  public Term get(Term key) {
    for (var map = this; map != EMPTY; map = map.parent) if (map.key.equals(key)) return map.value;
    return null;
  }

  public int size() {
    var n = 0;
    for (var map = this; map != EMPTY; map = map.parent) n++;
    return n;
  }
}

File: tools/derivation/FMap.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace derivation
{
    public sealed class FMap
    {
        public static readonly FMap Empty = new FMap(null, null, null);
        readonly FMap parent;
        readonly Term key;
        readonly Term value;

        FMap() { }

        FMap(FMap parent, Term key, Term value)
        {
            this.parent = parent;
            this.key = key;
            this.value = value;
        }

        public FMap Add(Term key, Term value)
        {
            return new FMap(this, key, value);
        }

        public Term this[Term key]
        {
            get
            {
                for (var map = this; map != Empty; map = map.parent)
                    if (map.key == key)
                        return map.value;
                return null;
            }
        }
    }
}
